frontend/packages/console-shared/src/components/contextMenu/PopupKebabMenu.tsx

```diff
@@ -0,0 +1,174 @@
+import * as React from 'react';
+import * as classNames from 'classnames';
+import * as _ from 'lodash';
+import { DropdownMenu } from '@patternfly/react-core';
+import { history, KebabItem, KebabOption } from '@console/internal/components/utils';
+import './PopupKebabMenu.scss';
+
+export interface PopupKebabMenuProps {
+  kebabOptions: KebabOption[];
+  className?: string;
+  onClose?(): void;
+  eventX: number;
+  eventY: number;
+  container?: Element;
+}
+interface PopopKebabMenuState {
+  menuTop: number;
+  menuLeft: number;
+}
+
+const MENU_PADDING = 20;
+
+export class PopupKebabMenu extends React.Component<PopupKebabMenuProps, PopopKebabMenuState> {
+  state = {
+    menuTop: 0,
+    menuLeft: 0,
+  };
+
+  backdrop = null;
+
+  menu = null;
+
+  componentDidMount() {
+    this.menu.addEventListener('mousedown', this.handleMenuMouseDown);
+    this.backdrop.addEventListener('mousedown', this.handleClose);
+    document.addEventListener('focus', this.handleBlur, true);
+    this.updateMenuPosition();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (
+      this.props.eventX !== prevProps.eventX ||
+      this.props.eventY !== prevProps.eventY ||
+      !_.isEqual(this.props.kebabOptions, prevProps.kebabOptions)
+    ) {
+      this.updateMenuPosition();
+    }
+  }
+
+  componentWillUnmount() {
+    this.menu.removeEventListener('mousedown', this.handleMenuMouseDown);
+    this.backdrop.removeEventListener('mousedown', this.handleClose);
+    document.removeEventListener('focus', this.handleBlur, true);
+  }
+
+  updateMenuPosition = () => {
+    const { container, eventX, eventY } = this.props;
+    let menuTop = eventY;
+    let menuLeft = eventX;
+
+    if (container) {
+      const menu = this.menu.firstChild;
+      if (menu) {
+        const elementRect = container.getBoundingClientRect();
+        const menuRect = menu.getBoundingClientRect();
+
+        if (
+          eventY + menuRect.height + menuRect.top >
+          elementRect.top + elementRect.height - MENU_PADDING
+        ) {
+          menuTop = Math.max(
+            elementRect.top + elementRect.height - menuRect.height - menuRect.top - MENU_PADDING,
+            0,
+          );
+        }
+        if (
+          eventX + menuRect.width + menuRect.left >
+          elementRect.left + elementRect.width - MENU_PADDING
+        ) {
+          menuLeft = Math.max(
+            elementRect.left + elementRect.width - menuRect.width - menuRect.left - MENU_PADDING,
+            0,
+          );
+        }
+      }
+    }
+
+    this.setState({ menuTop, menuLeft });
```

Comment:
Don't update state if the `menuTop` and `menuLeft` are the same since this isn't a `PureComponent` and this is called from `componentDidUpdate`.